### PR TITLE
drivers/lcd: code deduplication for st7735 and ili9341

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

In preparation for the parallel interface support the following changes were made:

1. The code for basic communication (acquire/release SPI device, SPI transfers), which were always duplicated by copy & paste in each display driver again and again, has been moved to the LCD driver as low-level functions that are now used by the display drivers. These low level function allow 
- code deduplication and
- to define a more abstract communication interface that can then be extended later by parallel communication interface
2. Identical GPIO initializations has also been moved from display drivers to the LCD driver.
3. Using a default implementation of `lcd_set_area` function allows further code deduplication.

Finally, the `ili9341` and `st7735` drivers only implement the inialization sequence for the LCD driver IC.

### Testing procedure

`tests/drivers/ili9341` should still work for a board with an ILI9341. Tested with `esp32-wrover-kit`.
`tests/drivers/st7735` should still work for a board with a ST77xx. Tested with `esp32s3-usb-otg`.

### CODE REVIEW

File: drivers/lcd/lcd.c [hunk 1]
```diff
@@ -25,45 +27,163 @@
...
+static inline void _lcd_write_byte(const lcd_t *dev, bool cont, uint8_t data)
 {
-    spi_acquire(dev->params->spi, dev->params->cs_pin, dev->params->spi_mode,
-                dev->params->spi_clk);
+    if (dev->params->spi != SPI_UNDEF) {
+        /* SPI serial interface is used */
+        spi_transfer_byte(dev->params->spi, dev->params->cs_pin, cont, data);
+    }
+    else {
+        assert(false);
```
Comment (aabadie) [inline drivers/lcd/lcd.c:45]: Why not just use `assert(dev->params->spi != SPI_UNDEF);` at the beginning of the function? (Here and below)

File: drivers/st7735/Kconfig [hunk 2]
```diff
@@ -22,19 +22,19 @@ config HAVE_ST7735
     help
       Indicates that an ST7735 display is present.
 
+config HAVE_ST7789
```
Comment (aabadie) [inline drivers/st7735/Kconfig:25]: This seems unrelated and actually not needed. The pattern here was to group driver module variants  module with their `HAS_VARIANT` just after. With that change it looks messed up.

Maybe a better approach is to use:

```kconfig
config MODULE_VARIANT1
    ...

config MODULE_VARIANT2
    ...

config HAS_VARIANT1
    ...

config HAS_VARIANT2
    ...
```

File: drivers/lcd/lcd.c (hunk 1 above)
Comment (gschorcht) [reply drivers/lcd/lcd.c:45]: Yeah, the reason is that this change is a split-off from parallel interface implementation where it will be
```c
#if IS_USED(MODULE_LCD_SPI)
    if (dev->params->spi != SPI_UNDEF) {
        /* SPI serial interface is used */
        spi_transfer_byte(dev->params->spi, dev->params->cs_pin, cont, data);
    }
    else {
#endif
#if IS_USED(MODULE_LCD_PARALLEL)
        /* MCU 8080 8-/16-bit parallel interface is used */
        _lcd_write_byte_par(dev, cont, data);
#else
        assert(false);
#endif
#if IS_USED(MODULE_LCD_SPI)
    }
#endif
```

I can change it, if you want :smile:

File: drivers/st7735/Kconfig (hunk 2 above)
Comment (gschorcht) [reply drivers/st7735/Kconfig:25]: It is from the history. When @maribu tried to fix master compilation in CI, he added
```diff
  config MODULE_ST7789
-     bool "ST7789 display driver"
-     select MODULE_ST7735    bool
+     depends on HAVE_ST7789
+     default y if MODULE_ST7735
+     help
        ST7789 display driver
```
according to my suggestion. This was the commit cb0aba6669183fc403910d3dc578a9bb01ca746f in this PR. When I rebased, the order has been changed.

Regarding the order, my feeling was that we define a `HAVE_*` symbol before it is used in `MODULE_*`. I can change it if you want.

File: drivers/lcd/lcd.c (hunk 1 above)
Comment (aabadie) [reply drivers/lcd/lcd.c:45]: Let's keep it like this if it's easier for you to rebase once that's merged.

File: drivers/st7735/Kconfig (hunk 2 above)
Comment (gschorcht) [reply drivers/st7735/Kconfig:32]: I have changed the order.

Why is this fix not correct? `MODULE_ST7789` is only available if the board enables `HAVE_ST7789`. The `MODULE_ST7789` has to be enabled by default if `MODULE_ST7735` is enabled. This corresponds to makefile dependencies.

**Review by aabadie — APPROVED** [review]
This is just code moved around and will simplify the work for coming PRs on new LCD drivers (not only SPI based). I saw some screenshot in privates that prove the PR is working.

Please squash!

ACK

File: drivers/st7735/Kconfig (hunk 2 above)
Comment (aabadie) [reply drivers/st7735/Kconfig:32]: "HAVE_ST7759" indeed tells the board has it so it can be pulled in automatically. But I manually plug the device to a dev kit and want to pull in myself, I'll also have to pullin the HAVE_ symbol which is not needed for other drivers AFAICT.

Comment (gschorcht) [reply drivers/st7735/Kconfig:32]: > But I manually plug the device to a dev kit and want to pull in myself, I'll also have to pullin the HAVE_ symbol

Yes, indeed. Unfortunatly, there was no other a approach to enable the `st7789` variant if the board pulls in the module in makefile dependencies. This should become obsolte once the driver varaiants are splitted.

**Comment by gschorcht** [comment]
@aabadie Thanks for reviewing and merging.
